File: .claude/typecheck-on-edit.py
#!/usr/bin/env python3
# Sanity check: mise x -- uvx ty check .claude/typecheck-on-edit.py
# and: mise x -- uvx ruff check .claude/typecheck-on-edit.py
import json
import os
import re
import subprocess
import sys
from pathlib import Path

data = json.load(sys.stdin)
path = data.get("tool_input", {}).get("file_path", "")
project_root = os.environ.get("CLAUDE_PROJECT_DIR", ".")

if not re.search(r"\.(tsx?|mts|cts)$", path):
    sys.exit(0)

match = re.search(r"(?:^|/)(packages|apps)/([^/]+)/", path)
tsconfig = (
    f"{match.group(1)}/{match.group(2)}/tsconfig.json" if match else "tsconfig.json"
)

if not Path(project_root, tsconfig).is_file():
    sys.exit(0)

result = subprocess.run(
    ["bunx", "tsc", "--noEmit", "-p", tsconfig],
    cwd=project_root,
    capture_output=True,
    text=True,
    check=False,
)
if result.returncode != 0:
    sys.stderr.write(
        f"tsc --noEmit -p {tsconfig} failed:\n{result.stdout}{result.stderr}",
    )
    sys.exit(2)

File: .agents/hooks/typecheck-on-edit.ts
#!/usr/bin/env bun
// Sanity check: bun run typecheck (tsc --noEmit, .agents/hooks is in root tsconfig.json include)

import { join } from "node:path";
import { readHookInput } from "./claude-hook";

const TYPECHECKED_FILE_PATTERN = /\.(tsx?|mts|cts)$/;
const PACKAGE_SCOPE_PATTERN = /(?:^|\/)(packages|apps)\/([^/]+)\//;

const input = await readHookInput();
const path = input.tool_input?.file_path ?? "";
const projectRoot = process.env.CLAUDE_PROJECT_DIR ?? ".";

if (!TYPECHECKED_FILE_PATTERN.test(path)) {
  process.exit(0);
}

const scopeMatch = path.match(PACKAGE_SCOPE_PATTERN);
const tsconfig = scopeMatch
  ? `${scopeMatch[1]}/${scopeMatch[2]}/tsconfig.json`
  : "tsconfig.json";

if (!(await Bun.file(join(projectRoot, tsconfig)).exists())) {
  process.exit(0);
}

const proc = Bun.spawn(["bunx", "tsc", "--noEmit", "-p", tsconfig], {
  cwd: projectRoot,
  stdout: "pipe",
  stderr: "pipe",
});
const [stdout, stderr, exitCode] = await Promise.all([
  new Response(proc.stdout).text(),
  new Response(proc.stderr).text(),
  proc.exited,
]);
if (exitCode !== 0) {
  process.stderr.write(
    `tsc --noEmit -p ${tsconfig} failed:\n${stdout}${stderr}`,
  );
  process.exit(2);
}
